leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java

```diff
@@ -261,76 +158,112 @@ protected Object handleInvocation(Object proxy, Method method, Object[] args) th
             }
         }
-
-        return leadershipToken;
-    }
-
-    private boolean isStillCurrentToken(LeadershipToken leadershipToken) {
-        return leadershipTokenRef.get() == leadershipToken;
-    }
-
-    private NotCurrentLeaderException notCurrentLeaderException(String message, @Nullable Throwable cause) {
-        return leaderElectionService
-                .getRecentlyPingedLeaderHost()
-                .map(hostAndPort -> new NotCurrentLeaderException(message, cause, hostAndPort))
-                .orElseGet(() -> new NotCurrentLeaderException(message, cause));
     }
 
-    private NotCurrentLeaderException notCurrentLeaderException(String message) {
-        return notCurrentLeaderException(message, null /* cause */);
+    private void claimResourcesOnLeadershipUpdate() {
+        maybeValidLeadershipTokenRef.set(null);
+        clearDelegate();
     }
 
-    private void markAsNotLeading(final LeadershipToken leadershipToken, @Nullable Throwable cause) {
-        log.warn("Lost leadership", cause);
-        if (leadershipTokenRef.compareAndSet(leadershipToken, null)) {
-            // this is fine in the case that this node has been elected leader again (i.e. with a different leadership
-            // token). `onGainedLeadership` guarantees that the delegate will be refreshed *before* we get a new
-            // leadershipToken. We're closing here instead of relying on the close in `onGainedLeadership` to reclaim
-            // resources and does not affect correctness.
-
-            // if we were to move this above or below the CAS, we could race with `onGainedLeadership` and end up
-            // clearing `delegateRef`.
-            try {
-                clearDelegate();
-            } catch (Throwable t) {
-                // If close fails we should still try to gain leadership
-            }
-            tryToGainLeadership();
+    /**
+     * Right now there is no way to release resources quickly. In the event of loss of leadership, we wait till the
+     * LeadershipCoordinator has updated its state. Then, the next request can release the delegateRef in
+     * `getLeadershipToken`.
+     */
+    private void handleNotLeading(final LeadershipToken leadershipToken, @Nullable Throwable cause) {
+        if (maybeValidLeadershipTokenRef.compareAndSet(leadershipToken, null)) {
+            // We are not clearing delegateTokenRef here. This is fine as we are relying on `getLeadershipToken` to
+            // claim the resources for the next request if this node loses leadership.
+            // If this node gains leadership again (i.e. with a different leadership token),
+            // `tryToUpdateLeadershipToken` guarantees that the delegate will be refreshed *before* we get a new
+            // leadershipToken.
+            // If we were to clearDelegateRef here or outside the CAS, we could race with
+            // `tryToUpdateLeadershipToken` and end up clearing `delegateRef`.
+
```

Comment:
So the reason why we cannot clear the delegate here is that some other ALP could mark the coordinator as not leading which would kick off trying to gain leadership?

Reply:
Right, in case where we had leadership - then lost - then gained leadership again, if some other ALP forces LeadershipCoordinator to refresh its leadershipToken, then, 2 request threads in a different ALP instance could potentially race if -> one thread realized the loss of leadership and tries to clear the delegate and the other one realized that leadershipToken was refreshed and tries to refresh delegate in `tryToUpdateLeadershipToken`.